[Excel] (Custom functions) "Advanced Options" articles (#834)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-parameter-options.md

      zipCode = "98052";
  }

  if (dayOfWeek === undefined) {
    dayOfWeek = "Wednesday";
  }

  // Get weather report for specified zipCode and dayOfWeek.
  // ...
}
```

## Range parameters

Your custom function may accept a range of data as an input parameter. A function can also return a range of data. In JavaScript, a range of data is represented as a two-dimensional array.

For example, suppose that your function returns the second highest value from a range of numbers stored in Excel. The following function accepts the parameter `values`, which is of type `Excel.CustomFunctionDimensionality.matrix`. Note that in the JSON metadata for this function, the parameter's `type` property is set to `matrix`.

```js
function secondHighest(values){
  let highest = values[0][0], secondHighest = values[0][0];
  for(var i = 0; i < values.length; i++){
    for(var j = 1; j < values[i].length; j++){
      if(values[i][j] >= highest){
        secondHighest = highest;
        highest = values[i][j];
      }
      else if(values[i][j] >= secondHighest){
        secondHighest = values[i][j];
      }
    }
  }
  return secondHighest;
}
```

## Invocation context parameter

> Every custom function is automatically passed an `invocation` argument as the last argument, which can be used when you wish to find the address of a cell or handle what happens when [canceling a function](custom-functions-web-reqs.md#stream-and-cancel-functions). Even if you declare no parameters, your custom function has this parameter. This argument doesn't appear for a user in Excel. If you want to use `invocation` in your custom function, declare it as the last parameter.

Reviewer: ```suggestion
Every custom function is automatically passed an `invocation` argument as the last argument, which can be used to find the address of a cell or handle what happens when [canceling a function](custom-functions-web-reqs.md#stream-and-cancel-functions). Even if you declare no parameters, your custom function has this parameter. This argument doesn't appear for a user in Excel. If you want to use `invocation` in your custom function, declare it as the last parameter.
```

---
ms.date: 04/29/2019
description: Learn how to use different parameters within your custom functions, such as Excel ranges, optional parameters, invocation context, and more.   
title: Options for Excel custom functions (preview)
localization_priority: Normal
---

# Custom functions parameter options

Custom functions are configurable with many different options for parameters: 
- [Optional parameters](#custom-functions-optional-parameters)
- [Range parameters](#range-parameters)
- [Invocation context parameter](#invocation-context-parameter)

## Custom functions optional parameters

Whereas regular parameters are required, optional parameters are not. When a user invokes a function in Excel, optional parameters appear in brackets. In the following sample, the add function can optionally add a third number. This function appears as `=CONTOSO.ADD(first, second, [third])` in Excel.

```js
/**
 * Add two numbers
 * @customfunction 
 * @param {number} first First number
 * @param {number} second Second number
 * @param {number} [third] Third number to add. If omitted, third = 0.
 * @returns {number} The sum of the two numbers.
 */
function add(first, second, third) {

>   if (third) {

Author: Oh right!

      zipCode = "98052";
  }

  if (dayOfWeek === undefined) {
    dayOfWeek = "Wednesday";
  }

  // Get weather report for specified zipCode and dayOfWeek.
  // ...
}
```

## Range parameters

Your custom function may accept a range of data as an input parameter. A function can also return a range of data. In JavaScript, a range of data is represented as a two-dimensional array.

For example, suppose that your function returns the second highest value from a range of numbers stored in Excel. The following function accepts the parameter `values`, which is of type `Excel.CustomFunctionDimensionality.matrix`. Note that in the JSON metadata for this function, the parameter's `type` property is set to `matrix`.

```js
function secondHighest(values){
  let highest = values[0][0], secondHighest = values[0][0];
  for(var i = 0; i < values.length; i++){
    for(var j = 1; j < values[i].length; j++){
      if(values[i][j] >= highest){
        secondHighest = highest;
        highest = values[i][j];
      }
      else if(values[i][j] >= secondHighest){
        secondHighest = values[i][j];
      }
    }
  }
  return secondHighest;
}
```

## Invocation context parameter

> Every custom function is automatically passed an `invocation` argument as the last argument, which can be used to find the address of a cell or handle what happens when [canceling a function](custom-functions-web-reqs.md#stream-and-cancel-functions). Even if you declare no parameters, your custom function has this parameter. This argument doesn't appear for a user in Excel. If you want to use `invocation` in your custom function, declare it as the last parameter.

Author: 'invocation' is the d.ts name for now but we can suggest eng change it post //build

In the following code sample, the `invocation` context is explicitly stated for your reference.

```js
/**
 * Add two numbers

>  * @customfunction 

Reviewer 2: We should also show the TypeScript version because it is not easy to lookup which type to use for the invocation parameter. Similar to the following...
```typescript
function cellAddress(context: CustomFunctions.Invocation) {
  return context.address;
}
```

Author: Is this something we want to do with a toggle between the two or two samples? I thought we were trying to avoid putting in TS in docs in general but I agree it's a lot easier to look at in TS.

 * @param {number} first First number
 * @param {number} second Second number
 * @returns {number} The sum of the two (or optionally three) numbers.
 */
function add(first, second, invocation) {
  return first + second;
}
CustomFunctions.associate("ADD", add);
```

The parameter allows you to get the context of the invoking cell, which can be helpful in some scenarios including [discovering the address of a cell which invoke a custom function](#addressing-cells-context-parameter).

### Addressing cell's context parameter

In some cases you need to get the address of the cell that invoked your custom function. This is useful in the following types of scenarios:

> - Formatting ranges: Use the cell's address as the key to store information in [AsyncStorage](/office/dev/add-ins/excel/custom-functions-runtime#storing-and-accessing-data). Then, use [onCalculated](/javascript/api/excel/excel.worksheet#oncalculated) in Excel to load the key from `AsyncStorage`.

Reviewer 2: Need to change AsyncStorage -> Storage object.

- Displaying cached values: If your function is used offline, display stored cached values from `AsyncStorage` using `onCalculated`.
- Reconciliation: Use the cell's address to discover an origin cell to help you reconcile where processing is occurring.

To request an addressing cell's context in a function, you need to use a helper function to find the cell's address, such as the one in the following example. The information about a cell's address is exposed only if `@requiresAddress` is tagged in the function's comments.

> ```js

Reviewer 2: This isn't a helper function, it's an actual custom function...

            "type": "number",
            "dimensionality": "scalar"
        },
        {
            "name": "second",
            "description": "second number to add",
            "type": "number",
            "dimensionality": "scalar",
        },
        {
            "name": "third",
            "description": "third optional number to add",
            "type": "number",
            "dimensionality": "scalar",
            "optional": true
        }
    ]
}
```

When you define a function that contains one or more optional parameters, you should specify what happens when the optional parameters are undefined. In the following example, `zipCode` and `dayOfWeek` are both optional parameters for the `getWeatherReport` function. If the `zipCode` parameter is undefined, the default value is set to 98052. If the `dayOfWeek` parameter is undefined, it is set to Wednesday.

```js
function getWeatherReport(zipCode, dayOfWeek)
{
  if (zipCode === undefined) {
      zipCode = "98052";
  }

  if (dayOfWeek === undefined) {
    dayOfWeek = "Wednesday";
  }

  // Get weather report for specified zipCode and dayOfWeek
  // ...
}
```

## Range parameters

> Your custom function may accept a range of data as an input parameter, or it may return a range of data. In JavaScript, a range of data is represented as a two-dimensional array.

Reviewer 2: This indicates you can return a range of data but I don't see an explanation or code sample of how to do that. We should add that as well.

For example, suppose that your function returns the second highest value from a range of numbers stored in Excel. The following function accepts the parameter `values`, which is of type `Excel.CustomFunctionDimensionality.matrix`. Note that in the JSON metadata for this function, you would set the parameter's `type` property to `matrix`.

```js

> function secondHighest(values){

Reviewer 2: I tested this code with the following var and got 5 as a result (which is incorrect).
```javascript
var v = [
    [5,7],
    [9,2]
    ];
```

  let highest = values[0][0], secondHighest = values[0][0];
  for(var i = 0; i < values.length; i++){

>     for(var j = 1; j < values[i].length; j++){

Author: I *think* this is because highest will eventually be i = 0 in this nested thing and so the next loop checks for second highest (aka, highest excluding the thing in position 0)?

      if(values[i][j] >= highest){
        secondHighest = highest;
        highest = values[i][j];
      }
      else if(values[i][j] >= secondHighest){
        secondHighest = values[i][j];
      }
    }
  }
  return secondHighest;
}
```

## Invocation context parameter

> Every custom function automatically contains an `invocationContext` argument, which is invisibly present at the end of any list of parameters. Even if you declare no parameters, your custom function will have this parameter. This argument is not explicitly stated, so it does not appear to the end user in Excel, nor is it present in the code in your script files.

Reviewer 2: ```suggestion
Every custom function is automatically passed an `invocationContext` argument as the last argument. Even if you declare no parameters, your custom function will have this parameter. This argument does not appear for a user in Excel. If you want to use `invocationContext` in your custom function, declare it as the last parameter.
```

Author: ```suggestion
Every custom function is automatically passed an `invocationContext` argument as the last argument. Even if you declare no parameters, your custom function will have this parameter. This argument doesn't appear for a user in Excel. If you want to use `invocationContext` in your custom function, declare it as the last parameter.
```

In the following code sample, you'll see the invocationContext explicitly stated for your reference.

```JavaScript
subtract(one, two, invocationContext) {

>   return one - two;

Reviewer 2: Might be interesting to at least console.log a property from the context. Doesn't it have a few properties pertaining to your function like cancelable, or streaming?

}
```

The parameter allows you to get the context of the invoking cell, which can be helpful in some scenarios including [discovering the address of a cell which invoke a custom function](#addressing-cells-context-parameter).

## Addressing cell's context parameter

In some cases you'll need to get the address of the cell that invoked your custom function. This may be useful in the following types of scenarios:

- Formatting ranges: Use the cell's address as the key to store information in [AsyncStorage](https://docs.microsoft.com/office/dev/add-ins/excel/custom-functions-runtime#storing-and-accessing-data). Then, use [onCalculated](https://docs.microsoft.com/javascript/api/excel/excel.worksheet#oncalculated) in Excel to load the key from `AsyncStorage`.
- Displaying cached values: If your function is used offline, display stored cached values from `AsyncStorage` using `onCalculated`.
- Reconciliation: Use the cell's address to discover an origin cell to help you reconcile where processing is occurring.

> The information about a cell's address is exposed only if `requiresAddress` is marked as `true` in the function's JSON metadata file. The following sample gives an example of this:

Reviewer 3: Consider adding the second sentence's content to the first as a comma clause (", as shown in the following code sample.")

```JSON
{
   "id": "ADDTIME",
   "name": "ADDTIME",
   "description": "Display current date and add the amount of hours to it designated by the parameter",
   "helpUrl": "http://www.contoso.com",
   "result": {
      "type": "number",
      "dimensionality": "scalar"
   },
   "parameters": [
      {
         "name": "Additional time",
         "description": "Amount of hours to increase current date by",
         "type": "number",
         "dimensionality": "scalar"
      }
   ],
   "options": {
      "requiresAddress": true
   }
}
```

> In the script file (**./src/customfunctions.js** or **./src/customfunctions.ts**), you'll also need to add a `getAddress` function to find a cell's address. This function may take parameters, as shown in the following sample as `parameter1`. The last parameter will always be `invocationContext`, an object containing the cell's location that Excel passes down when `requiresAddress` is marked as `true` in your JSON metadata file.

Reviewer 3: Consider "you'll" -> "you"

> In the script file (**./src/customfunctions.js** or **./src/customfunctions.ts**), you'll also need to add a `getAddress` function to find a cell's address. This function may take parameters, as shown in the following sample as `parameter1`. The last parameter will always be `invocationContext`, an object containing the cell's location that Excel passes down when `requiresAddress` is marked as `true` in your JSON metadata file.

Reviewer 2: ```suggestion
In the script file (**./src/functions/functions.js** or **./src/functions/functions.ts**), you'll also need to add a `getAddress` function to find a cell's address. This function may take parameters, as shown in the following sample as `parameter1`. The last parameter will always be `invocationContext`, an object containing the cell's location that Excel passes down when `requiresAddress` is marked as `true` in your JSON metadata file.
```

```js

> function getAddress(parameter1, invocationContext) {

Reviewer 3: Is this function explicitly necessary? I'm a little confused, since this seems like a helper function, but that doesn't seem necessary if all functions get the invocationContext and we've ensured it has an address through the JSON.

Reviewer 2: I'm also confused if this is a helper function or a custom function.
